### PR TITLE
ddl: scatter the regions of table when creating them

### PR DESCRIPTION
**Pull request by jackysp**
Signed-off-by: Shuaipeng Yu <jackysp@gmail.com>

### What problem does this PR solve? 
It’s a bit late to schedule the region when PD found the hotspot. It is better to scatter them when they are empty.

### What is changed and how it works?
Scatter the regions of the tables when creating them.

### Check List 

Tests 

 - Manual test (add detailed scripts or steps below)
Created 100 tables and compare the regions distribution.
master:
屏幕快照 2019-06-28 下午7 11 10
This PR:
屏幕快照 2019-06-29 下午10 51 37

Code changes

 - Has exported variable/fields change

Side effects

 - Increased code complexity

### CODE REVIEW

**Comment by jackysp** [comment]
For 100 tables, waiting to scatter finish costs 24 min, and no waiting cost 6 min.

**Comment by jackysp** [comment]
Change `tidb_wait_split_region_finish` to global scope. It is a global only variable, so it keeps the same behavior as the global only variables in MySQL.
```
$ mysql -uroot test
Reading table information for completion of table and column names
You can turn off this feature to get a quicker startup with -A

Welcome to the MySQL monitor.  Commands end with ; or \g.
Your MySQL connection id is 15
Server version: 8.0.16 Homebrew

Copyright (c) 2000, 2019, Oracle and/or its affiliates. All rights reserved.

Oracle is a registered trademark of Oracle Corporation and/or its
affiliates. Other names may be trademarks of their respective
owners.

Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

mysql> load data local infile "~/tmp.csv" into table t;
Query OK, 1 row affected (0.07 sec)
Records: 1  Deleted: 0  Skipped: 0  Warnings: 0

mysql> set global local_infile=0;
Query OK, 0 rows affected (0.00 sec)

mysql> load data local infile "~/tmp.csv" into table t;
ERROR 1148 (42000): The used command is not allowed with this MySQL version
mysql> 
```

File: store/tikv/split_region.go [hunk 1]
```diff
@@ -29,12 +28,7 @@ import (
 
 // SplitRegion splits the region contains splitKey into 2 regions: [start,
 // splitKey) and [splitKey, end).
-func (s *tikvStore) SplitRegion(splitKey kv.Key) error {
-	_, err := s.splitRegion(splitKey)
-	return err
-}
-
-func (s *tikvStore) splitRegion(splitKey kv.Key) (*metapb.Region, error) {
+func (s *tikvStore) SplitRegion(splitKey kv.Key, scatter bool) (uint64, error) {
```
Comment (coocood) [inline store/tikv/split_region.go]: Need comment for the returned `uint64` value.

File: sessionctx/variable/sysvar.go [hunk 2]
```diff
@@ -697,7 +697,7 @@ var defaultSysVars = []*SysVar{
 	{ScopeGlobal | ScopeSession, TiDBOptJoinReorderThreshold, strconv.Itoa(DefTiDBOptJoinReorderThreshold)},
 	{ScopeSession, TiDBCheckMb4ValueInUTF8, BoolToIntStr(config.GetGlobalConfig().CheckMb4ValueInUTF8)},
 	{ScopeSession, TiDBSlowQueryFile, ""},
-	{ScopeSession, TiDBWaitSplitRegionFinish, BoolToIntStr(DefTiDBWaitSplitRegionFinish)},
+	{ScopeGlobal, TiDBWaitSplitRegionFinish, BoolToIntStr(DefTiDBWaitSplitRegionFinish)},
```
Comment (crazycs520) [inline sessionctx/variable/sysvar.go]: Why change to `Global` scope?

File: store/tikv/split_region.go (hunk 1 above)
Comment (winkyao) [reply store/tikv/split_region.go]: I think `SplitRegion` is better just used for splitting region, scatter region is another action. Use a boolean to indicate different actions is not clean code. Keep `SplitRegionAndScatter` is better?

Comment (jackysp) [reply store/tikv/split_region.go]: `SplitRegionAndScatter` is even bad. I think it’s worse to combine the names of the actions together as the function name. And we also need an interface named `SplitRegion`. The best way is to add a standalone interface for scattering, but now we have at least one interface saved, and split and scatter are really strongly related. If we need to use scatter separately in the future, I think we can separate them again.

**Comment by jackysp** [comment]
PTAL @crazycs520 @winkyao @coocood

**Comment by coocood** [comment]
LGTM

File: ddl/split_region.go [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+			if err != nil {
+				logutil.BgLogger().Warn("[ddl] pre split table region failed", zap.Int64("recordID", recordID), zap.Error(err))
+			} else {
+				regionIDs = append(regionIDs, regionID)
+			}
+		}
+	} else {
+		regionIDs = append(regionIDs, splitRecordRegion(store, tbInfo.ID, scatter))
+	}
+	regionIDs = append(regionIDs, splitIndexRegion(store, tbInfo, scatter)...)
```
[the remark this answers is not in the document]
Comment (jackysp) [reply ddl/split_region.go]: emm... Yes, it is expected.

**Comment by crazycs520** [comment]
@winkyao PTAL

**Comment by jackysp** [comment]
PTAL @winkyao

File: ddl/ddl_api.go [hunk 4]
```diff
@@ -1323,25 +1324,25 @@ func (d *ddl) CreateTable(ctx sessionctx.Context, s *ast.CreateTableStmt) (err e
...
 		// do pre-split and scatter.
-		if tbInfo.ShardRowIDBits > 0 && tbInfo.PreSplitRegions > 0 {
-			preSplitAndScatter = func() {
-				preSplitTableShardRowIDBitsRegion(d.store, tbInfo, ctx.GetSessionVars().WaitSplitRegionFinish)
+		sp, ok := d.store.(kv.SplitableStore)
+		if ok && atomic.LoadUint32(&EnableSplitTableRegion) != 0 {
+			var preSplit func()
+			val, err := variable.GetGlobalSystemVar(ctx.GetSessionVars(), variable.TiDBScatterRegion)
+			if err != nil {
+				logutil.BgLogger().Warn("[ddl] pre-split region failed", zap.Error(err))
```
Comment (winkyao) [inline ddl/ddl_api.go]: why not return?

File: ddl/ddl_api.go [hunk 5]
```diff
@@ -1323,25 +1324,25 @@ func (d *ddl) CreateTable(ctx sessionctx.Context, s *ast.CreateTableStmt) (err e
...
 			}
-		} else if atomic.LoadUint32(&EnableSplitTableRegion) != 0 {
+			scatterRegion := variable.TiDBOptOn(val)
```
Comment (winkyao) [inline ddl/ddl_api.go]: if the `variable.GetGlobalSystemVar` return err, scatterRegion should be false?

Comment (jackysp) [reply ddl/ddl_api.go]: Yes, I think so.

**Comment by crazycs520** [comment]
@jackysp Please cherry-pick this PR to v3.0 and v2.1.

**Pull request merged**